### ISSUE TITLE
Question about the encoder of mdt_transformer.py

### ISSUE THREAD

**Issue by jjihwan** [intuitive-robots/mdt_policy#16]
Hi, thanks a lot for your great work!

In [this line](https://github.com/intuitive-robots/mdt_policy/blob/e7a78e744751724d3faa5d113fe8cb5f678c052a/mdt/models/networks/mdt_transformer.py#L211)  of `mdt.models.networks.mdt_transformer.py`, it seems that the encoder of the transformer get action as an input.

However, what I've understood from your paper is that the encoder get only the current state and the goal as inputs, as Figure1 in your paper.
(I think the implementation of `mdt.models.networks.mdtv_transformer.py` is correct.)
Is the implementation wrong? or are there anything that I'm misunderstanding?

Looking forward your response!

**Comment by jjihwan** [comment]
Also, why the decoder of the transformer have 'causal attention', even though the decoder does not generate actions in an autoregressive way? I believe that bidirectional attention would enhance the performance.

**Comment by mbreuss** [comment]
> In [this line](https://github.com/intuitive-robots/mdt_policy/blob/e7a78e744751724d3faa5d113fe8cb5f678c052a/mdt/models/networks/mdt_transformer.py#L211) of mdt.models.networks.mdt_transformer.py, it seems that the encoder of the transformer get action as an input.

Good catch. So this is not intended and was a mistake. Sorry  for the inconvenience! 

> I believe that bidirectional attention would enhance the performance.

So this is an interesting problem without a definite answer. Some papers do same setup as I do, while others claim non-causal masking is better. I tried multiple times with different benchmarks and settings and causal version was better most of the times. So that's why I sticked with this causal masking. Let me know if you get any benefits from it :)

Just for reference since this code is already 1 year old and some parts of the model are outdated: don't use the Voltron or resent-18 from scratch. Using ResNet-50 with Imagenet pretraining as in [MoDE](https://github.com/intuitive-robots/MoDE_Diffusion_Policy) or my recommended VLM from [FLOWER](https://github.com/intuitive-robots/flower_vla_calvin) is way better. 

Let me know if you have any other questions!

**Comment by jjihwan** [comment]
Thank you for your prompt response and kind suggestion!

I'm just using your pretrained model to construct dataset for training my own video generator, and I just asked for clarification.
Your guidance was truly helpful for me, thanks!!

**Comment by jjihwan** [comment]
BTW it is quite interesting that causal model works better than non-causal model.
I think causal masking must result in restriction of the information in diffusion models.
Do you have any intuition or thought about the reason why causal model works better?

**Comment by mbreuss** [comment]
> I think causal masking must result in restriction of the information in diffusion models.

So I have one assumption: using the causal masking makes it harder to learn compared to bi-directional attention. Since DP overfit very well, maybe that helps in these smaller-scale benchmark settings. 

Recent generalist VLAs like Pi 0.5 use full attention. So maybe this is something regarding the data scale, where more data alleviates this issue. 

<img width="859" alt="Image" src="https://github.com/user-attachments/assets/54265969-b3ca-475a-ba33-b4b1f30160aa" />
